pkg/reconciler/inmemorychannel/dispatcher/inmemorychannel.go

```diff
@@ -209,11 +209,31 @@ func (r *Reconciler) patchSubscriberStatus(ctx context.Context, imc *v1.InMemory
 }
 
 // newConfigForInMemoryChannel creates a new Config for a single inmemory channel.
-func newConfigForInMemoryChannel(imc *v1.InMemoryChannel) (*multichannelfanout.ChannelConfig, error) {
+func newConfigForInMemoryChannel(ctx context.Context, imc *v1.InMemoryChannel) (*multichannelfanout.ChannelConfig, error) {
+	featureFlags := feature.FromContext(ctx)
+	if featureFlags.IsOIDCAuthentication() && len(imc.Spec.Subscribers) != len(imc.Status.Subscribers) {
+		return nil, fmt.Errorf("the number of subscribers in the status does not match the number of subscribers in the spec")
+	}
 	subs := make([]fanout.Subscription, len(imc.Spec.Subscribers))
 
+	// we will need to find the subscriber status for the service account, so let's make it easy to search for that
+	subStats := make(map[types.UID]eventingduckv1.SubscriberStatus, len(imc.Status.Subscribers))
+	for _, sub := range imc.Status.Subscribers {
+		subStats[sub.UID] = sub
+	}
+
 	for i, sub := range imc.Spec.Subscribers {
 		conf, err := fanout.SubscriberSpecToFanoutConfig(sub)
+		if featureFlags.IsOIDCAuthentication() {
+			if subStat, ok := subStats[sub.UID]; ok {
+				conf.ServiceAccount = &types.NamespacedName{
+					Name:      *subStat.Auth.ServiceAccountName,
```

Comment:
I don't think we have the auth status at this point yet:
```
  subscribers:
  - observedGeneration: 3
    ready: "True"
    uid: 9c5dec83-6849-46e4-9dca-a8a8bdf1829e
```

https://github.com/knative/eventing/blob/4f864c7da3e44d8e36a81a780363bc9dab8fa879/pkg/reconciler/inmemorychannel/dispatcher/inmemorychannel.go#L183-L187

Probably need an additional field in the SubscriberSpec: https://github.com/knative/eventing/blob/7a645f87cc1ec798a8171ef8de16d936de2036db/pkg/apis/duck/v1/subscribable_types.go#L37-L71